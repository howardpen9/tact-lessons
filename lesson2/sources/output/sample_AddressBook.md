# TACT Compilation Report
Contract: AddressBook
BOC Size: 2403 bytes

# Types
Total Types: 10

## StateInit
TLB: `_ code:^cell data:^cell = StateInit`
Signature: `StateInit{code:^cell,data:^cell}`

## Context
TLB: `_ bounced:bool sender:address value:int257 raw:^slice = Context`
Signature: `Context{bounced:bool,sender:address,value:int257,raw:^slice}`

## SendParameters
TLB: `_ bounce:bool to:address value:int257 mode:int257 body:Maybe ^cell code:Maybe ^cell data:Maybe ^cell = SendParameters`
Signature: `SendParameters{bounce:bool,to:address,value:int257,mode:int257,body:Maybe ^cell,code:Maybe ^cell,data:Maybe ^cell}`

## ChangeOwner
TLB: `change_owner#819dbe99 queryId:uint64 newOwner:address = ChangeOwner`
Signature: `ChangeOwner{queryId:uint64,newOwner:address}`

## ChangeOwnerOk
TLB: `change_owner_ok#327b2b4a queryId:uint64 newOwner:address = ChangeOwnerOk`
Signature: `ChangeOwnerOk{queryId:uint64,newOwner:address}`

## AddUserMessage
TLB: `add_user_message#945f9817 name:^string address:address = AddUserMessage`
Signature: `AddUserMessage{name:^string,address:address}`

## RemoveUserMessage
TLB: `remove_user_message#2568a530 name:^string = RemoveUserMessage`
Signature: `RemoveUserMessage{name:^string}`

## ProxyMessage
TLB: `proxy_message#82a4a809 to:^string message:^string = ProxyMessage`
Signature: `ProxyMessage{to:^string,message:^string}`

## User
TLB: `_ name:^string address:address received_count:uint32 sent_count:uint32 = User`
Signature: `User{name:^string,address:address,received_count:uint32,sent_count:uint32}`

## Name
TLB: `_ name:^string = Name`
Signature: `Name{name:^string}`

# Get Methods
Total Get Methods: 3

## get_stat
Argument: str

## get_username
Argument: address

## owner

# Error Codes
2: Stack undeflow
3: Stack overflow
4: Integer overflow
5: Integer out of expected range
6: Invalid opcode
7: Type check error
8: Cell overflow
9: Cell underflow
10: Dictionary error
13: Out of gas error
32: Method ID not found
34: Action is invalid or not supported
37: Not enough TON
38: Not enough extra-currencies
128: Null reference exception
129: Invalid serialization prefix
130: Invalid incoming message
131: Constraints error
132: Access denied
133: Contract stopped
134: Invalid argument
135: Code of a contract was not found
136: Invalid address
137: Masterchain support is not enabled for this contract