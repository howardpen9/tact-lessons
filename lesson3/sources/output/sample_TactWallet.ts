import { 
    Cell,
    Slice, 
    Address, 
    Builder, 
    beginCell, 
    ComputeError, 
    TupleItem, 
    TupleReader, 
    Dictionary, 
    contractAddress, 
    ContractProvider, 
    Sender, 
    Contract, 
    ContractABI, 
    TupleBuilder,
    DictionaryValue
} from 'ton-core';

export type StateInit = {
    $$type: 'StateInit';
    code: Cell;
    data: Cell;
}

export function storeStateInit(src: StateInit) {
    return (builder: Builder) => {
        let b_0 = builder;
        b_0.storeRef(src.code);
        b_0.storeRef(src.data);
    };
}

export function loadStateInit(slice: Slice) {
    let sc_0 = slice;
    let _code = sc_0.loadRef();
    let _data = sc_0.loadRef();
    return { $$type: 'StateInit' as const, code: _code, data: _data };
}

function loadTupleStateInit(source: TupleReader) {
    let _code = source.readCell();
    let _data = source.readCell();
    return { $$type: 'StateInit' as const, code: _code, data: _data };
}

function storeTupleStateInit(source: StateInit) {
    let builder = new TupleBuilder();
    builder.writeCell(source.code);
    builder.writeCell(source.data);
    return builder.build();
}

function dictValueParserStateInit(): DictionaryValue<StateInit> {
    return {
        serialize: (src, buidler) => {
            buidler.storeRef(beginCell().store(storeStateInit(src)).endCell());
        },
        parse: (src) => {
            return loadStateInit(src.loadRef().beginParse());
        }
    }
}

export type Context = {
    $$type: 'Context';
    bounced: boolean;
    sender: Address;
    value: bigint;
    raw: Cell;
}

export function storeContext(src: Context) {
    return (builder: Builder) => {
        let b_0 = builder;
        b_0.storeBit(src.bounced);
        b_0.storeAddress(src.sender);
        b_0.storeInt(src.value, 257);
        b_0.storeRef(src.raw);
    };
}

export function loadContext(slice: Slice) {
    let sc_0 = slice;
    let _bounced = sc_0.loadBit();
    let _sender = sc_0.loadAddress();
    let _value = sc_0.loadIntBig(257);
    let _raw = sc_0.loadRef();
    return { $$type: 'Context' as const, bounced: _bounced, sender: _sender, value: _value, raw: _raw };
}

function loadTupleContext(source: TupleReader) {
    let _bounced = source.readBoolean();
    let _sender = source.readAddress();
    let _value = source.readBigNumber();
    let _raw = source.readCell();
    return { $$type: 'Context' as const, bounced: _bounced, sender: _sender, value: _value, raw: _raw };
}

function storeTupleContext(source: Context) {
    let builder = new TupleBuilder();
    builder.writeBoolean(source.bounced);
    builder.writeAddress(source.sender);
    builder.writeNumber(source.value);
    builder.writeSlice(source.raw);
    return builder.build();
}

function dictValueParserContext(): DictionaryValue<Context> {
    return {
        serialize: (src, buidler) => {
            buidler.storeRef(beginCell().store(storeContext(src)).endCell());
        },
        parse: (src) => {
            return loadContext(src.loadRef().beginParse());
        }
    }
}

export type SendParameters = {
    $$type: 'SendParameters';
    bounce: boolean;
    to: Address;
    value: bigint;
    mode: bigint;
    body: Cell | null;
    code: Cell | null;
    data: Cell | null;
}

export function storeSendParameters(src: SendParameters) {
    return (builder: Builder) => {
        let b_0 = builder;
        b_0.storeBit(src.bounce);
        b_0.storeAddress(src.to);
        b_0.storeInt(src.value, 257);
        b_0.storeInt(src.mode, 257);
        if (src.body !== null && src.body !== undefined) { b_0.storeBit(true).storeRef(src.body); } else { b_0.storeBit(false); }
        if (src.code !== null && src.code !== undefined) { b_0.storeBit(true).storeRef(src.code); } else { b_0.storeBit(false); }
        if (src.data !== null && src.data !== undefined) { b_0.storeBit(true).storeRef(src.data); } else { b_0.storeBit(false); }
    };
}

export function loadSendParameters(slice: Slice) {
    let sc_0 = slice;
    let _bounce = sc_0.loadBit();
    let _to = sc_0.loadAddress();
    let _value = sc_0.loadIntBig(257);
    let _mode = sc_0.loadIntBig(257);
    let _body = sc_0.loadBit() ? sc_0.loadRef() : null;
    let _code = sc_0.loadBit() ? sc_0.loadRef() : null;
    let _data = sc_0.loadBit() ? sc_0.loadRef() : null;
    return { $$type: 'SendParameters' as const, bounce: _bounce, to: _to, value: _value, mode: _mode, body: _body, code: _code, data: _data };
}

function loadTupleSendParameters(source: TupleReader) {
    let _bounce = source.readBoolean();
    let _to = source.readAddress();
    let _value = source.readBigNumber();
    let _mode = source.readBigNumber();
    let _body = source.readCellOpt();
    let _code = source.readCellOpt();
    let _data = source.readCellOpt();
    return { $$type: 'SendParameters' as const, bounce: _bounce, to: _to, value: _value, mode: _mode, body: _body, code: _code, data: _data };
}

function storeTupleSendParameters(source: SendParameters) {
    let builder = new TupleBuilder();
    builder.writeBoolean(source.bounce);
    builder.writeAddress(source.to);
    builder.writeNumber(source.value);
    builder.writeNumber(source.mode);
    builder.writeCell(source.body);
    builder.writeCell(source.code);
    builder.writeCell(source.data);
    return builder.build();
}

function dictValueParserSendParameters(): DictionaryValue<SendParameters> {
    return {
        serialize: (src, buidler) => {
            buidler.storeRef(beginCell().store(storeSendParameters(src)).endCell());
        },
        parse: (src) => {
            return loadSendParameters(src.loadRef().beginParse());
        }
    }
}

export type Deploy = {
    $$type: 'Deploy';
    queryId: bigint;
}

export function storeDeploy(src: Deploy) {
    return (builder: Builder) => {
        let b_0 = builder;
        b_0.storeUint(2490013878, 32);
        b_0.storeUint(src.queryId, 64);
    };
}

export function loadDeploy(slice: Slice) {
    let sc_0 = slice;
    if (sc_0.loadUint(32) !== 2490013878) { throw Error('Invalid prefix'); }
    let _queryId = sc_0.loadUintBig(64);
    return { $$type: 'Deploy' as const, queryId: _queryId };
}

function loadTupleDeploy(source: TupleReader) {
    let _queryId = source.readBigNumber();
    return { $$type: 'Deploy' as const, queryId: _queryId };
}

function storeTupleDeploy(source: Deploy) {
    let builder = new TupleBuilder();
    builder.writeNumber(source.queryId);
    return builder.build();
}

function dictValueParserDeploy(): DictionaryValue<Deploy> {
    return {
        serialize: (src, buidler) => {
            buidler.storeRef(beginCell().store(storeDeploy(src)).endCell());
        },
        parse: (src) => {
            return loadDeploy(src.loadRef().beginParse());
        }
    }
}

export type DeployOk = {
    $$type: 'DeployOk';
    queryId: bigint;
}

export function storeDeployOk(src: DeployOk) {
    return (builder: Builder) => {
        let b_0 = builder;
        b_0.storeUint(2952335191, 32);
        b_0.storeUint(src.queryId, 64);
    };
}

export function loadDeployOk(slice: Slice) {
    let sc_0 = slice;
    if (sc_0.loadUint(32) !== 2952335191) { throw Error('Invalid prefix'); }
    let _queryId = sc_0.loadUintBig(64);
    return { $$type: 'DeployOk' as const, queryId: _queryId };
}

function loadTupleDeployOk(source: TupleReader) {
    let _queryId = source.readBigNumber();
    return { $$type: 'DeployOk' as const, queryId: _queryId };
}

function storeTupleDeployOk(source: DeployOk) {
    let builder = new TupleBuilder();
    builder.writeNumber(source.queryId);
    return builder.build();
}

function dictValueParserDeployOk(): DictionaryValue<DeployOk> {
    return {
        serialize: (src, buidler) => {
            buidler.storeRef(beginCell().store(storeDeployOk(src)).endCell());
        },
        parse: (src) => {
            return loadDeployOk(src.loadRef().beginParse());
        }
    }
}

export type FactoryDeploy = {
    $$type: 'FactoryDeploy';
    queryId: bigint;
    cashback: Address;
}

export function storeFactoryDeploy(src: FactoryDeploy) {
    return (builder: Builder) => {
        let b_0 = builder;
        b_0.storeUint(1829761339, 32);
        b_0.storeUint(src.queryId, 64);
        b_0.storeAddress(src.cashback);
    };
}

export function loadFactoryDeploy(slice: Slice) {
    let sc_0 = slice;
    if (sc_0.loadUint(32) !== 1829761339) { throw Error('Invalid prefix'); }
    let _queryId = sc_0.loadUintBig(64);
    let _cashback = sc_0.loadAddress();
    return { $$type: 'FactoryDeploy' as const, queryId: _queryId, cashback: _cashback };
}

function loadTupleFactoryDeploy(source: TupleReader) {
    let _queryId = source.readBigNumber();
    let _cashback = source.readAddress();
    return { $$type: 'FactoryDeploy' as const, queryId: _queryId, cashback: _cashback };
}

function storeTupleFactoryDeploy(source: FactoryDeploy) {
    let builder = new TupleBuilder();
    builder.writeNumber(source.queryId);
    builder.writeAddress(source.cashback);
    return builder.build();
}

function dictValueParserFactoryDeploy(): DictionaryValue<FactoryDeploy> {
    return {
        serialize: (src, buidler) => {
            buidler.storeRef(beginCell().store(storeFactoryDeploy(src)).endCell());
        },
        parse: (src) => {
            return loadFactoryDeploy(src.loadRef().beginParse());
        }
    }
}

export type ExtMessage = {
    $$type: 'ExtMessage';
    signature: Buffer;
    seqno: bigint;
    valid_until: bigint;
    message_parameters: SendParameters;
}

export function storeExtMessage(src: ExtMessage) {
    return (builder: Builder) => {
        let b_0 = builder;
        b_0.storeUint(3240438462, 32);
        b_0.storeBuffer(src.signature);
        b_0.storeUint(src.seqno, 32);
        b_0.storeUint(src.valid_until, 32);
        let b_1 = new Builder();
        b_1.store(storeSendParameters(src.message_parameters));
        b_0.storeRef(b_1.endCell());
    };
}

export function loadExtMessage(slice: Slice) {
    let sc_0 = slice;
    if (sc_0.loadUint(32) !== 3240438462) { throw Error('Invalid prefix'); }
    let _signature = sc_0.loadBuffer(64);
    let _seqno = sc_0.loadUintBig(32);
    let _valid_until = sc_0.loadUintBig(32);
    let sc_1 = sc_0.loadRef().beginParse();
    let _message_parameters = loadSendParameters(sc_1);
    return { $$type: 'ExtMessage' as const, signature: _signature, seqno: _seqno, valid_until: _valid_until, message_parameters: _message_parameters };
}

function loadTupleExtMessage(source: TupleReader) {
    let _signature = source.readBuffer();
    let _seqno = source.readBigNumber();
    let _valid_until = source.readBigNumber();
    const _message_parameters = loadTupleSendParameters(source.readTuple());
    return { $$type: 'ExtMessage' as const, signature: _signature, seqno: _seqno, valid_until: _valid_until, message_parameters: _message_parameters };
}

function storeTupleExtMessage(source: ExtMessage) {
    let builder = new TupleBuilder();
    builder.writeBuffer(source.signature);
    builder.writeNumber(source.seqno);
    builder.writeNumber(source.valid_until);
    builder.writeTuple(storeTupleSendParameters(source.message_parameters));
    return builder.build();
}

function dictValueParserExtMessage(): DictionaryValue<ExtMessage> {
    return {
        serialize: (src, buidler) => {
            buidler.storeRef(beginCell().store(storeExtMessage(src)).endCell());
        },
        parse: (src) => {
            return loadExtMessage(src.loadRef().beginParse());
        }
    }
}

 type TactWallet_init_args = {
    $$type: 'TactWallet_init_args';
    publicKey: bigint;
}

function initTactWallet_init_args(src: TactWallet_init_args) {
    return (builder: Builder) => {
        let b_0 = builder;
        b_0.storeInt(src.publicKey, 257);
    };
}

async function TactWallet_init(publicKey: bigint) {
    const __code = Cell.fromBase64('te6ccgECGgEAA/UAART/APSkE/S88sgLAQIBIAIDAgFIBAUCMvLbPFjbPDDI+EMBzH8BygBZAssfy//J7VQSEwKc0AHQ0wMBcbCjAfpAASDXSYEBC7ry4Igg1wsKIIEE/7ry0ImDCbry4IhUUFMDbwT4YQL4Yts8Wts88uCCyPhDAcx/AcoAWQLLH8v/ye1UEgYCASAICQHW7aLt+wGSMH/gcCHXScIflTAg1wsf3iDAACLXScEhsJJbf+AgghCUapi2uo6oMNMfAYIQlGqYtrry4IHTPwExyAGCEK/5D1dYyx/LP8n4QgFwbds8f+DAAJ8g10nCH5iAINchMH/bMeDeMH8HATptbSJus5lbIG7y0IBvIgGRMuIQJHADBIBCUCPbPBgCASAKCwIBIA4PAhG7Oc2zzbPGwhgSDAIRuMl9s82zxsIYEg0AAiAAAiEAlbu9GCcFzsPV0srnsehOw51kqFG2aCcJ3WNS0rZHyzItOvLf3xYjmCcCBVwBuAZ2OUzlg6rkclssOCcJ2XTlqzTstzOg6WbZRm6KSAIBSBARABGwr7tRNDSAAGAAdbJu40NWlwZnM6Ly9RbVE3OVdhejhUSm1HWXJCckt5TW03ZExhb0JtMzFGUkw4THFmUXhZWFlHcHplggAVDtRNDUAfhj0gABl9Mf0/9ZbBLg+CjXCwqDCbry4ImBAQHXAAEB0ds8FAEwcCHXScIflTAg1wsf3oIQwSUqvrrjAjBwFQAEcAED1NMfAYIQwSUqvrry4IGDCNcY0x/TH9QB0Ns8NxB6EHkQeFUFbBooyMsfUoDLH1R3ZVR3ZSfIVWDbPMkBzMn5AIElaVG8+RAa8vSBdyxRi7oY8vSBLwH4I1AHuxby9PgACKQQNkUYREPbPH8WFxgAktIA+kABINdJgQELuvLgiCDXCwoggQT/uvLQiYMJuvLgiAGBAQHXAIEBAdcA0gABkdSSbQHi0gABkdSSbQHi0gABkdSSbQHiVWAAuFBnygBQBCDXSYEBC7ry4Igg1wsKIIEE/7ry0ImDCbry4IjPFhKBAQHPAIEBAc8AIW6zlX8BygDMlHAyygDiIW6zlX8BygDMlHAyygDiIW6zlX8BygDMlHAyygDiAcrIcQHKAVAHAcoAcAHKAlAFINdJgQELuvLgiCDXCwoggQT/uvLQiYMJuvLgiM8WUAP6AnABymgjbrORf5MkbrPilzMzAXABygDjDSFus5x/AcoAASBu8tCAAcyVMXABygDiyQH7ABkAmH8BygDIcAHKAHABygAkbrOdfwHKAAQgbvLQgFAEzJY0A3ABygDiJG6znX8BygAEIG7y0IBQBMyWNANwAcoA4nABygACfwHKAALJWMw=');
    const __system = Cell.fromBase64('te6cckECHAEAA/8AAQHAAQEFoMNXAgEU/wD0pBP0vPLICwMCASAJBAIy8ts8WNs8MMj4QwHMfwHKAFkCyx/L/8ntVBoFATBwIddJwh+VMCDXCx/eghDBJSq+uuMCMHAGA9TTHwGCEMElKr668uCBgwjXGNMf0x/UAdDbPDcQehB5EHhVBWwaKMjLH1KAyx9Ud2VUd2UnyFVg2zzJAczJ+QCBJWlRvPkQGvL0gXcsUYu6GPL0gS8B+CNQB7sW8vT4AAikEDZFGERD2zx/CAcYALhQZ8oAUAQg10mBAQu68uCIINcLCiCBBP+68tCJgwm68uCIzxYSgQEBzwCBAQHPACFus5V/AcoAzJRwMsoA4iFus5V/AcoAzJRwMsoA4iFus5V/AcoAzJRwMsoA4gCS0gD6QAEg10mBAQu68uCIINcLCiCBBP+68tCJgwm68uCIAYEBAdcAgQEB1wDSAAGR1JJtAeLSAAGR1JJtAeLSAAGR1JJtAeJVYAIBSBUKAgEgEAsCASAPDAIBSA4NAHWybuNDVpcGZzOi8vUW1RNzlXYXo4VEptR1lyQnJLeU1tN2RMYW9CbTMxRlJMOExxZlF4WVhZR3B6ZYIAARsK+7UTQ0gABgAJW7vRgnBc7D1dLK57HoTsOdZKhRtmgnCd1jUtK2R8syLTry398WI5gnAgVcAbgGdjlM5YOq5HJbLDgnCdl05as07LczoOlm2UZuikgCASATEQIRuMl9s82zxsIYGhIAAiECEbs5zbPNs8bCGBoUAAIgApzQAdDTAwFxsKMB+kABINdJgQELuvLgiCDXCwoggQT/uvLQiYMJuvLgiFRQUwNvBPhhAvhi2zxa2zzy4ILI+EMBzH8BygBZAssfy//J7VQaFgHW7aLt+wGSMH/gcCHXScIflTAg1wsf3iDAACLXScEhsJJbf+AgghCUapi2uo6oMNMfAYIQlGqYtrry4IHTPwExyAGCEK/5D1dYyx/LP8n4QgFwbds8f+DAAJ8g10nCH5iAINchMH/bMeDeMH8XATptbSJus5lbIG7y0IBvIgGRMuIQJHADBIBCUCPbPBgByshxAcoBUAcBygBwAcoCUAUg10mBAQu68uCIINcLCiCBBP+68tCJgwm68uCIzxZQA/oCcAHKaCNus5F/kyRus+KXMzMBcAHKAOMNIW6znH8BygABIG7y0IABzJUxcAHKAOLJAfsAGQCYfwHKAMhwAcoAcAHKACRus51/AcoABCBu8tCAUATMljQDcAHKAOIkbrOdfwHKAAQgbvLQgFAEzJY0A3ABygDicAHKAAJ/AcoAAslYzAFQ7UTQ1AH4Y9IAAZfTH9P/WWwS4Pgo1wsKgwm68uCJgQEB1wABAdHbPBsABHABP7ncvg==');
    let builder = beginCell();
    builder.storeRef(__system);
    builder.storeUint(0, 1);
    initTactWallet_init_args({ $$type: 'TactWallet_init_args', publicKey })(builder);
    const __data = builder.endCell();
    return { code: __code, data: __data };
}

const TactWallet_errors: { [key: number]: { message: string } } = {
    2: { message: `Stack undeflow` },
    3: { message: `Stack overflow` },
    4: { message: `Integer overflow` },
    5: { message: `Integer out of expected range` },
    6: { message: `Invalid opcode` },
    7: { message: `Type check error` },
    8: { message: `Cell overflow` },
    9: { message: `Cell underflow` },
    10: { message: `Dictionary error` },
    13: { message: `Out of gas error` },
    32: { message: `Method ID not found` },
    34: { message: `Action is invalid or not supported` },
    37: { message: `Not enough TON` },
    38: { message: `Not enough extra-currencies` },
    128: { message: `Null reference exception` },
    129: { message: `Invalid serialization prefix` },
    130: { message: `Invalid incoming message` },
    131: { message: `Constraints error` },
    132: { message: `Access denied` },
    133: { message: `Contract stopped` },
    134: { message: `Invalid argument` },
    135: { message: `Code of a contract was not found` },
    136: { message: `Invalid address` },
    137: { message: `Masterchain support is not enabled for this contract` },
    9577: { message: `Invalid Signature` },
    12033: { message: `Invalid Time` },
    30508: { message: `Invalid Seqno` },
}

export class TactWallet implements Contract {
    
    static async init(publicKey: bigint) {
        return await TactWallet_init(publicKey);
    }
    
    static async fromInit(publicKey: bigint) {
        const init = await TactWallet_init(publicKey);
        const address = contractAddress(0, init);
        return new TactWallet(address, init);
    }
    
    static fromAddress(address: Address) {
        return new TactWallet(address);
    }
    
    readonly address: Address; 
    readonly init?: { code: Cell, data: Cell };
    readonly abi: ContractABI = {
        errors: TactWallet_errors
    };
    
    private constructor(address: Address, init?: { code: Cell, data: Cell }) {
        this.address = address;
        this.init = init;
    }
    
    async send(provider: ContractProvider, via: Sender, args: { value: bigint, bounce?: boolean| null | undefined }, message: null | string | Slice | Deploy) {
        
        let body: Cell | null = null;
        if (message === null) {
            body = new Cell();
        }
        if (typeof message === 'string') {
            body = beginCell().storeUint(0, 32).storeStringTail(message).endCell();
        }
        if (message && typeof message === 'object' && message instanceof Slice) {
            body = message.asCell();
        }
        if (message && typeof message === 'object' && !(message instanceof Slice) && message.$$type === 'Deploy') {
            body = beginCell().store(storeDeploy(message)).endCell();
        }
        if (body === null) { throw new Error('Invalid message type'); }
        
        await provider.internal(via, { ...args, body: body });
        
    }
    
    async sendExternal(provider: ContractProvider, message: ExtMessage) {
        
        let body: Cell | null = null;
        if (message && typeof message === 'object' && !(message instanceof Slice) && message.$$type === 'ExtMessage') {
            body = beginCell().store(storeExtMessage(message)).endCell();
        }
        if (body === null) { throw new Error('Invalid message type'); }
        
        await provider.external(body);
        
    }
    
    async getSeqno(provider: ContractProvider) {
        let builder = new TupleBuilder();
        let source = (await provider.get('seqno', builder.build())).stack;
        let result = source.readBigNumber();
        return result;
    }
    
    async getGetPublicKey(provider: ContractProvider) {
        let builder = new TupleBuilder();
        let source = (await provider.get('get_public_key', builder.build())).stack;
        let result = source.readBigNumber();
        return result;
    }
    
}