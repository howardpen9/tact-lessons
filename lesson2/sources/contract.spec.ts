import { toNano } from "ton";
import { ContractSystem } from "@tact-lang/emulator";
import { AddressBook } from "./output/sample_AddressBook";

describe("contract", () => {
    it("should deploy correctly", async () => {
        // Create ContractSystem and deploy contract
        let system = await ContractSystem.create();
        let owner = system.treasure("owner");
        let someone = system.treasure("someone");
        let contract = system.open(await AddressBook.fromInit(owner.address));

        let contract_tracker = system.track(contract.address);
        
        await contract.send(owner, {value: toNano("0.1")}, {$$type: 'AddUserMessage', name: "owner", address: owner.address});
        await contract.send(owner, {value: toNano("0.1")}, {$$type: 'AddUserMessage', name: 'someone', address: someone.address});
        await system.run();

        await contract.send(owner, {value: toNano("0.1")}, {$$type: 'ProxyMessage', message: "Hello", to: "someone"});
        await contract.send(someone, {value: toNano("0.1")}, {$$type: 'ProxyMessage', message: "Hello", to: "owner"});
        await contract.send(someone, {value: toNano("0.1")}, {$$type: 'ProxyMessage', message: "Hello", to: "owner"});
        await system.run();
        await contract.send(someone, {value: toNano("0.1")}, "someone");
        await system.run();

        expect(contract_tracker.collect()).toMatchSnapshot();
    });
});
