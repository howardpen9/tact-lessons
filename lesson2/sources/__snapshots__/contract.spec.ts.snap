// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contract should deploy correctly 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{945F981780011F6E292AB9FE5825731C5F568DB15E68F69707390F1976DA6A7F173916703B9_}
 x{6F776E6572}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(owner)",
          "to": "kQDLpyT8CJHookyURfvJlUMxGDXEqIPxQXYIM1a1s_jOqYxn",
          "type": "internal",
          "value": "0.1",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 9745n,
      },
    ],
  },
  {
    "$seq": 1,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{945F98178017D7106C5BAFA91E82BDC51DA9B2540F408179A12D717B9427AB9F178EF13BB89_}
 x{736F6D656F6E65}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(owner)",
          "to": "kQDLpyT8CJHookyURfvJlUMxGDXEqIPxQXYIM1a1s_jOqYxn",
          "type": "internal",
          "value": "0.1",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 11716n,
      },
    ],
  },
  {
    "$seq": 2,
    "events": [
      {
        "$type": "storage-charged",
        "amount": "0.000000013",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{82A4A809}
 x{736F6D656F6E65}
 x{48656C6C6F}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(owner)",
          "to": "kQDLpyT8CJHookyURfvJlUMxGDXEqIPxQXYIM1a1s_jOqYxn",
          "type": "internal",
          "value": "0.1",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 20165n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "Hello",
              "type": "text",
            },
            "bounce": true,
            "from": "kQDLpyT8CJHookyURfvJlUMxGDXEqIPxQXYIM1a1s_jOqYxn",
            "to": "@treasure(someone)",
            "type": "internal",
            "value": "0.078663",
          },
        ],
      },
    ],
  },
  {
    "$seq": 3,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{82A4A809}
 x{6F776E6572}
 x{48656C6C6F}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(someone)",
          "to": "kQDLpyT8CJHookyURfvJlUMxGDXEqIPxQXYIM1a1s_jOqYxn",
          "type": "internal",
          "value": "0.1",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 20165n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "Hello",
              "type": "text",
            },
            "bounce": true,
            "from": "kQDLpyT8CJHookyURfvJlUMxGDXEqIPxQXYIM1a1s_jOqYxn",
            "to": "@treasure(owner)",
            "type": "internal",
            "value": "0.078663",
          },
        ],
      },
    ],
  },
  {
    "$seq": 4,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{82A4A809}
 x{6F776E6572}
 x{48656C6C6F}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(someone)",
          "to": "kQDLpyT8CJHookyURfvJlUMxGDXEqIPxQXYIM1a1s_jOqYxn",
          "type": "internal",
          "value": "0.1",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 20165n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "Hello",
              "type": "text",
            },
            "bounce": true,
            "from": "kQDLpyT8CJHookyURfvJlUMxGDXEqIPxQXYIM1a1s_jOqYxn",
            "to": "@treasure(owner)",
            "type": "internal",
            "value": "0.078663",
          },
        ],
      },
    ],
  },
  {
    "$seq": 5,
    "events": [
      {
        "$type": "storage-charged",
        "amount": "0.000000013",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "someone",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(someone)",
          "to": "kQDLpyT8CJHookyURfvJlUMxGDXEqIPxQXYIM1a1s_jOqYxn",
          "type": "internal",
          "value": "0.1",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 34270n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "someone sent 2 message(s) (66.66%) and received 1 message(s) (33.33%)",
              "type": "text",
            },
            "bounce": true,
            "from": "kQDLpyT8CJHookyURfvJlUMxGDXEqIPxQXYIM1a1s_jOqYxn",
            "to": "@treasure(someone)",
            "type": "internal",
            "value": "0.064046",
          },
        ],
      },
    ],
  },
]
`;
